torch/csrc/jit/autodiff.cpp

```diff
@@ -137,33 +138,48 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
     "aten::eq(Tensor self, Tensor other) -> Tensor",
     "aten::ne(Tensor self, Tensor other) -> Tensor"
   };
-  const auto build_sym_grad = [node](const std::vector<SymbolicVariable>& grads) -> std::vector<SymbolicVariable> {
+  const auto sumToSizeOf = [node](SymbolicVariable v, Symbol input_name) -> SymbolicVariable {
+    Value * size;
+    {
+      WithInsertPoint insert_guard {node};
+      size = SymbolicVariable(node->namedInput(input_name)).size();
+    }
+    return v.sumToSize(size);
+  };
+  const auto build_sym_grad = [node, &sumToSizeOf](const std::vector<SymbolicVariable>& grads) -> std::vector<SymbolicVariable> {
     auto inputs = fmap<SymbolicVariable>(node->inputs());
     auto outputs = fmap<SymbolicVariable>(node->outputs());
 
     if (node->matches("aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor")) {
-      return {grads.at(0), grads.at(0) * node->namedInput(attr::alpha), nullptr};
+      return {sumToSizeOf(grads.at(0), attr::self),
+              sumToSizeOf(grads.at(0) * node->namedInput(attr::alpha), attr::other),
+              nullptr};
 
     } else if (node->matches("aten::add(Tensor self, Scalar other, Scalar alpha) -> Tensor")) {
       return {grads.at(0), nullptr, nullptr};
 
     } else if (node->kind() == prim::AutogradAdd) {
+      // NB: AutogradAdds don't broadcast
       return {grads.at(0), grads.at(0)};
 
     } else if (node->matches("aten::sub(Tensor self, Tensor other, *, Scalar alpha) -> Tensor")) {
-      return {grads.at(0), -grads.at(0) * node->namedInput(attr::alpha), nullptr};
+      return {sumToSizeOf(grads.at(0), attr::self),
+              sumToSizeOf(-grads.at(0) * node->namedInput(attr::alpha), attr::other),
+              nullptr};
 
     } else if (node->matches("aten::sub(Tensor self, Scalar other, Scalar alpha) -> Tensor")) {
       return {grads.at(0), nullptr, nullptr};
 
     } else if (node->matches("aten::mul(Tensor self, Tensor other) -> Tensor")) {
-      return {grads.at(0) * inputs.at(1), grads.at(0) * inputs.at(0)};
+      return {sumToSizeOf(grads.at(0) * inputs.at(1), attr::self),
+              sumToSizeOf(grads.at(0) * inputs.at(0), attr::other)};
 
     } else if (node->matches("aten::mul(Tensor self, Scalar other) -> Tensor")) {
       return {grads.at(0) * inputs.at(1), nullptr};
 
     } else if (node->matches("aten::div(Tensor self, Tensor other) -> Tensor")) {
-      return {grads.at(0) / inputs.at(1), -grads.at(0) * inputs.at(0) / (inputs.at(1) * inputs.at(1))};
+      return {sumToSizeOf(grads.at(0) / inputs.at(1), attr::self),
```

Comment:
(putting this comment here because github won't let me comment where I want to)
max and min also broadcast, so those need `sumToSize` as well

Reply:
I thought max and min insert explicit expands https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/shape_analysis.cpp#L1382-L1383 and expands are handled in backward? I'm also wondering if it would be better to insert expands in forward where needed, rather than unconditionally put sumToSize everywhere?

Reply:
I'm not sure -- I don't see explicit expands in
```
import torch

def fnsum(a, b):
    a = a.unsqueeze(2)
    b = b.unsqueeze(1)
    sum = torch.max(a, b)
    out = torch.tanh(sum)
    return out

if __name__ == '__main__':
    a1 = torch.randn(2, 3, 5, requires_grad=True)
    a2 = torch.randn(2, 4, 5, requires_grad=True)
    fn = torch.jit.script(fnsum)
    out = fn(a1, a2)
    print(fn.graph_for(a1, a2))
    out.backward(torch.ones_like(out))
```

Reply:
@ngimel Script never inserts explicit expands.

Reply:
Ideally, there could be a middle ground between hard-coded sizes and promises like "there won't be a broadcast in this op" or "expect dim x of tensor y to be broadcasted". In many usecases, tensor sizes can be different, but dimensionality and contiguity stay the same.

Reply:
Dimensionality (in terms of the number of dimensions) and contiguity yes, but those are insufficient guarantees to prove that no broadcasts will happen.

Reply:
Symbolic expressions might be that happy middle ground!

Reply:
Fixed.